src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp

```diff
@@ -0,0 +1,227 @@
+   * @param nodeIndex The number of the node being deleted.
+   */
+  template<typename TreeType>
+  void RemoveNode(TreeType* node, const size_t nodeIndex);
+
+  /**
+   * Copy the largest Hilbert value and the local dataset
+   * @param dst The node to which the information is being copied.
+   * @param src The node from which the information is being copied.
+   */
+  template<typename TreeType>
+  void Copy(TreeType* dst, TreeType* src);
+  
+  void NullifyData();
+  
+  /**
+   * Update the largest Hilbert value and the local dataset.
+   * The children of the node (or the points that the node contains) should be
+   * arranged according to their Hilbert values.
+   * @param node The node in which the information should be updated.
+   */
+  template<typename TreeType>
+  void UpdateLargestValue(TreeType* node);
+
+  template<typename TreeType>
+  void UpdateHilbertValues(TreeType* parent, size_t firstSibling,
+                           size_t lastSibling);
+
+  /**
+   * Calculate the Hilbert value of the point pt.
+   * @param pt The point for which the Hilbert value should be calculated.
+   */
+  template<typename VecType>
+  static arma::Col<HilbertElemType> CalculateValue(const VecType& pt,
+                             typename boost::enable_if<IsVector<VecType>>* = 0);
+
+  /**
+   * Compare two Hilbert values. It returns 1 if the first value is greater than
+   * the second one, -1 if the first value is less than the second one and
+   * 0 if the values are equal. This method does not compute the Hilbert values.
+   * @param value1 The first value.
+   * @param value2 The second value.
+   */
+  static int CompareValues(const arma::Col<HilbertElemType>& value1,
+                           const arma::Col<HilbertElemType>& value2);
+
+  //! Return the number of values
+  size_t NumValues() const
+  { return numValues; }
+
+  //! Modify the number of values
+  size_t& NumValues()
+  { return numValues; }
+
+  //! Return the local dataset
+  const arma::Mat<HilbertElemType>* LocalDataset() const
+  { return localDataset; }
+
+  //! Modify the dataset
+  arma::Mat<HilbertElemType>*& LocalDataset() { return localDataset; }
+  
+  //! Modify the valueToInsert
+  arma::Col<HilbertElemType>* ValueToInsert() { return valueToInsert; }
+
+  //! Modify the valueToInsert
+  const arma::Col<HilbertElemType>* ValueToInsert() const
+  { return valueToInsert; }
+
+ private:
+  //! The number of bits that we can store
+  static constexpr size_t order = sizeof(HilbertElemType) * CHAR_BIT;
+  //! The local dataset
+  arma::Mat<HilbertElemType>* localDataset;
+  //! Indicates that the node owns the local dataset
+  bool ownsLocalDataset;
+  //! The number of values in the local dataset
+  size_t numValues;
+  //! The Hilbert value of the point that is being inserted
+  arma::Col<HilbertElemType>* valueToInsert;
+  //! Indicates that the node owns the valueToInsert 
+  bool ownsValueToInsert;
```

Comment:
I read the code more and maybe my comment is misguided, since the constructor for the `DiscreteHilbertValue` class will try to take the cached value of the parent node to set `valueToInsert`.  Maybe there is still a way to refactor here but I am not so sure, so maybe there is nothing to do here.